File: onto-code-maker-result-tester/src/main/java/ontology/generator/classes/examples/familySimple/serialization/PersonSerialization.java

package ontology.generator.classes.examples.familySimple.serialization;

import org.eclipse.rdf4j.model.*;
import java.util.Collection;
import java.util.Set;
import java.util.HashSet;
import java.util.*;

import java.util.stream.Collectors;
import ontology.generator.classes.examples.familySimple.entities.*;
import org.eclipse.rdf4j.model.util.Values;
import org.eclipse.rdf4j.model.vocabulary.RDF;

import java.time.LocalTime;
import ontology.generator.classes.examples.familySimple.Vocabulary;

public class PersonSerialization extends SerializationModel<Person>{

    @Override
    public void addToModel(Model model, Person person) {
        model.add(person.getIri(),RDF.TYPE, person.getClassIRI());
        addPropertiesToModel(model,person);

    }

    protected void addPropertiesToModel(Model model,  Person person) {
        if(person.getHasAge() != null){
            model.add(person.getIri(),Vocabulary.HASAGE_PROPERTY_IRI,Values.literal(person.getHasAge()));
        }

        List<OntoEntity> hasCatPom = new ArrayList<>();
        hasCatPom.addAll(person.getHasCat());
        setRDFCollection(model,person.getIri(),Vocabulary.HASCAT_PROPERTY_IRI,hasCatPom);

        if(person.getHasDog() != null ){
            model.add(person.getIri(),Vocabulary.HASDOG_PROPERTY_IRI,person.getHasDog().getIri());
        }



    }

    protected void setProperties(Model model,Person person,int nestingLevel) throws Exception{
        Literal hasAge = super.getFirstLiteralObject(model,Vocabulary.HASAGE_PROPERTY_IRI,person.getIri());
        if ( hasAge != null ){
            person.setHasAge(Integer.valueOf(hasAge.intValue()) );
        }
        Set<Resource> hasCat = super.getAllResourceObjects(model,Vocabulary.HASCAT_PROPERTY_IRI,person.getIri());
        for(Resource propValue:hasCat){
            if(propValue.isIRI()) {
                Cat hasCatInstance = new CatSerialization().getInstanceFromModel(model, (IRI) propValue,nestingLevel);
                if(hasCatInstance == null) throw new Exception("Instance of " + propValue.stringValue() + " is not in model.");
                person.addHasCat(hasCatInstance);
            }else if(propValue.isBNode()){
                List<Value> listOfValues = super.getRDFCollection(model,(BNode)propValue);
                for(Value value:listOfValues){
                    if(value.isIRI()){
                        Cat hasCatInstance = new CatSerialization().getInstanceFromModel(model, (IRI)value,nestingLevel);
                        if(hasCatInstance == null) throw new Exception("Instance of " + propValue.stringValue() + " is not in model.");
                        person.addHasCat(hasCatInstance);
                    }
                 }
            }
        }
        IRI hasDog = super.getFirstIriObject(model,Vocabulary.HASDOG_PROPERTY_IRI,person.getIri());
         if ( hasDog == null){
              // check equivalent property hasDogEq
               hasDog = super.getFirstIriObject(model,Vocabulary.HASDOGEQ_PROPERTY_IRI,person.getIri());
         }
        if ( hasDog != null ){
            Dog hasDogInstance = new DogSerialization().getInstanceFromModel(model, hasDog,nestingLevel);
            person.setHasDog(hasDogInstance);
        }

    }

    @Override
    public Person getInstanceFromModel(Model model,IRI instanceIri,int nestingLevel) throws Exception{
        Model statements = model.filter(instanceIri,RDF.TYPE,Person.CLASS_IRI);
        if(statements.size() != 0){
            Person person = new Person(instanceIri);
            if(nestingLevel > 0){
                nestingLevel--;
                setProperties(model, person,nestingLevel);
            }
            return person;
        }

        return null;
    }

    @Override
    public Collection<Person> getAllInstancesFromModel(Model model,int nestingLevel)throws Exception{
        Model statements = model.filter(null,RDF.TYPE,Person.CLASS_IRI);
        Collection<Person> allInstances = new ArrayList<>();
        for(Statement statement:statements){
            Resource subject = statement.getSubject();
            if(subject.isIRI()){
                IRI iri = (IRI) subject;
                Person person = getInstanceFromModel(model,iri,nestingLevel);
                allInstances.add(person);
            }
        }

        return allInstances;
    }

    @Override
    public void removeInstanceFromModel(Model model,IRI instanceIri) {
        Set<Value> rdfCollections = model.filter(instanceIri,null,null).objects();
        List<Value> listOfnodes = rdfCollections.stream().filter(o -> o.isBNode()).collect(Collectors.toList());
        for(Value node:listOfnodes){
            model.removeAll(getModelRDFCollection(model,(BNode)node));
            model.remove(instanceIri,null,(BNode)node);
        }
        model.remove(instanceIri,RDF.TYPE,Person.CLASS_IRI);
        model.remove(instanceIri,null,null);
    }

    @Override
    public void updateInstanceInModel(Model model,Person person){
        Set<Value> rdfCollections = model.filter(person.getIri(),null,null).objects();
        List<Value> listOfnodes = rdfCollections.stream().filter(Value::isBNode).collect(Collectors.toList());
        for(Value node:listOfnodes){
            model.removeAll(getModelRDFCollection(model,(BNode)node));
            model.remove(person.getIri(),null,node);
        }

        Model statements = model.filter(person.getIri(),null,null);
        statements.removeIf(x -> !x.getPredicate().equals(RDF.TYPE));

        addPropertiesToModel(model,person);
    }



}

File: onto-code-maker-result-tester/src/test/python/familyPython/serialization/PersonSerialization.py

from rdflib.graph import URIRef, Literal,BNode,Collection
from rdflib.namespace import RDF
from Vocabulary import Vocabulary
from entities.Person import Person
from .SerializationModel import SerializationModel

class PersonSerialization(SerializationModel):

    def addToModel(self,ontology, person):
        ontology.add((person.getIri(),RDF.type, person.getClassIRI()))
        self.addPropertiesToModel(ontology,person)


    def addPropertiesToModel(self,ontology, person):
        hasLuckyNumbersPom = set(person.getHasLuckyNumbers())
        super().setLiteralsRDFCollection(ontology,person.getIri(),Vocabulary.HASLUCKYNUMBERS_PROPERTY_IRI,hasLuckyNumbersPom)
        if person.getHasAge() is not None:
            ontology.add((person.getIri(),Vocabulary.HASAGE_PROPERTY_IRI,Literal(person.getHasAge())))

        hasCatPom = set(person.getHasCat())
        super().setRDFCollection(ontology,person.getIri(),Vocabulary.HASCAT_PROPERTY_IRI,hasCatPom)
        if person.getHasDog() is not None:
            ontology.add((person.getIri(),Vocabulary.HASDOG_PROPERTY_IRI,person.getHasDog().getIri()))

        agePom = set(person.getAge())
        super().setLiteralsRDFCollection(ontology,person.getIri(),Vocabulary.AGE_PROPERTY_IRI,agePom)
        pass

    def setProperties(self,ontology, person, nestingLevel):
        hasLuckyNumbers = super().getAllObjects(ontology,Vocabulary.HASLUCKYNUMBERS_PROPERTY_IRI,person.getIri())
        for propValue in hasLuckyNumbers:
            if isinstance(propValue,Literal):
                person.addHasLuckyNumbers(str(propValue))
            elif isinstance(propValue,BNode):
                listOfValues = super().getRDFCollection(ontology,propValue)
                for literalValue in listOfValues :
                    if isinstance(literalValue,Literal):
                        person.addHasLuckyNumbers(str(literalValue))

        hasAge = super().getFirstLiteralObject(ontology,Vocabulary.HASAGE_PROPERTY_IRI,person.getIri())
        if hasAge is not None:
            person.setHasAge(str(hasAge))
        hasCat = super().getAllResourceObjects(ontology,Vocabulary.HASCAT_PROPERTY_IRI,person.getIri())
        for propValue in hasCat:
            if isinstance(propValue,URIRef):
                from .CatSerialization import CatSerialization
                hasCatInstance = CatSerialization().getInstanceFromModel(ontology, propValue,nestingLevel)
                if hasCatInstance is None:
                    print("Instance of " + str(propValue) + " is not in model.")
                person.addHasCat(hasCatInstance)
            elif isinstance(propValue,BNode):
                listOfValues = super().getRDFCollection(ontology,propValue)
                for value in listOfValues:
                    if isinstance(value,URIRef):
                        from .CatSerialization import CatSerialization
                        hasCatInstance = CatSerialization().getInstanceFromModel(ontology,value,nestingLevel)
                        if hasCatInstance is None:
                            print("Instance of " + str(propValue) + " is not in model.")
                        person.addHasCat(hasCatInstance)
        hasDog = super().getFirstIriObject(ontology,Vocabulary.HASDOG_PROPERTY_IRI,person.getIri())
        if hasDog is None:
            # check equivalent property hasDogEq
            hasDog = super().getFirstIriObject(ontology,Vocabulary.HASDOGEQ_PROPERTY_IRI,person.getIri())

        if hasDog is not None:
            from .DogSerialization import DogSerialization
            hasDogInstance = DogSerialization().getInstanceFromModel(ontology, hasDog,nestingLevel)
            person.setHasDog(hasDogInstance)

        age = super().getAllObjects(ontology,Vocabulary.AGE_PROPERTY_IRI,person.getIri())
        for propValue in age:
            if isinstance(propValue,Literal):
                person.addAge(str(propValue))
            elif isinstance(propValue,BNode):
                listOfValues = super().getRDFCollection(ontology,propValue)
                for literalValue in listOfValues :
                    if isinstance(literalValue,Literal):
                        person.addAge(str(literalValue))

        pass

    def getInstanceFromModel(self,ontology, instanceIri,nestingLevel):
        if (instanceIri,RDF.type,Person.CLASS_IRI) in ontology:
            person =  Person(instanceIri)
            if nestingLevel > 0:
                nestingLevel -= 1
                self.setProperties(ontology, person,nestingLevel)
            return person
        return None

    def getAllInstancesFromModel(self,ontology,nestingLevel):
        allInstances = set()
        for s, p, o in ontology.triples((None,RDF.type,Person.CLASS_IRI)):
            if isinstance(s,URIRef) :
                person = self.getInstanceFromModel(ontology,s,nestingLevel)
                allInstances.add(person)
        return allInstances

    def removeInstanceFromModel(self,ontology, subject):
        ontology.remove((subject,RDF.type,Person.CLASS_IRI))
        for statement in ontology.triples((subject,None,None)):
            ontology.remove(statement)



